README.md
# Plug Logger JSON

[![Build Status](https://travis-ci.org/bleacherreport/plug_logger_json.svg?branch=master)](https://travis-ci.org/bleacherreport/plug_logger_json)
[![Module Version](https://img.shields.io/hexpm/v/plug_logger_json.svg)](https://hex.pm/packages/plug_logger_json)
[![Hex Docs](https://img.shields.io/badge/hex-docs-lightgreen.svg)](https://hexdocs.pm/plug_logger_json/)
[![Total Download](https://img.shields.io/hexpm/dt/plug_logger_json.svg)](https://hex.pm/packages/plug_logger_json)
[![License](https://img.shields.io/hexpm/l/plug_logger_json.svg)](https://github.com/bleacherreport/plug_logger_json/blob/master/LICENSE)
[![Last Updated](https://img.shields.io/github/last-commit/bleacherreport/plug_logger_json.svg)](https://github.com/bleacherreport/plug_logger_json/commits/master)


A comprehensive JSON logger Plug.

## Dependencies

* `Plug`
* `Poison`

## Elixir & Erlang Support

The support policy is to support the last 2 major versions of Erlang and the three last minor versions of Elixir.

## Installation

1. Add `:plug_logger_json` to your list of dependencies in `mix.exs`:

   ```elixir
   def deps do
    [
      {:plug_logger_json, "~> 0.7.0"}
    ]
   end
   ```

2. Ensure `:plug_logger_json` is started before your application (Skip if using Elixir 1.4 or greater):

   ```elixir
   def application do
     [applications: [:plug_logger_json]]
   end
   ```

3. Replace `Plug.Logger` with either:

   * `Plug.LoggerJSON, log: Logger.level`,
   * `Plug.LoggerJSON, log: Logger.level, extra_attributes_fn: &MyPlug.extra_attributes/1` in your plug pipeline (in `endpoint.ex` for Phoenix apps),

## Recommended Setup

### Configure `plug_logger_json`

Add to your `config/config.exs` or `config/env_name.exs` if you want to filter params or headers or suppress any logged keys:

```elixir
config :plug_logger_json,
  filtered_keys: ["password", "authorization"],
  suppressed_keys: ["api_version", "log_type"]
```

### Configure the logger (console)

In your `config/config.exs` or `config/env_name.exs`:

```elixir
config :logger, :console,
  format: "$message\n",
  level: :info, # You may want to make this an env variable to change verbosity of the logs
  metadata: [:request_id]
```

### Configure the logger (file)

Do the following:

* Update deps in `mix.exs` with the following:

  ```elixir
  def deps do
  [
    {:logger_file_backend, "~> 0.0.10"}
  ]
  end
  ```

* Add to your `config/config.exs` or `config/env_name.exs`:

  ```elixir
  config :logger,
    format: "$message\n",
    backends: [{LoggerFileBackend, :log_file}, :console]

  config :logger, :log_file,
    format: "$message\n",
    level: :info,
    metadata: [:request_id],
    path: "log/my_pipeline.log"
  ```

* Ensure you are using `Plug.Parsers` (Phoenix adds this to `endpoint.ex` by default) to parse params as well as request body:

  ```elixir
  plug Plug.Parsers,
    parsers: [:urlencoded, :multipart, :json],
    pass: ["*/*"],
    json_decoder: Poison
  ```

## Error Logging

In `router.ex` of your Phoenix project or in your plug pipeline:

* Add `require Logger`,
* Add `use Plug.ErrorHandler`,
* Add the following two private functions:

  ```elixir
  defp handle_errors(%Plug.Conn{status: 500} = conn, %{kind: kind, reason: reason, stack: stacktrace}) do
    Plug.LoggerJSON.log_error(kind, reason, stacktrace)
    send_resp(conn, 500, Poison.encode!(%{errors: %{detail: "Internal server error"}}))
  end

  defp handle_errors(_, _), do: nil
  ```

## Extra Attributes

Additional data can be logged alongside the request by specifying a function to call which returns a map:

```elixir
def extra_attributes(conn) do
  map = %{
    "user_id" => get_in(conn.assigns, [:user, :user_id]),
    "other_id" => get_in(conn.private, [:private_resource, :id]),
    "should_not_appear" => conn.private[:does_not_exist]
  }

  map
  |> Enum.filter(&(&1 !== nil))
  |> Enum.into(%{})
end

plug Plug.LoggerJSON,
  log: Logger.level(),
  extra_attributes_fn: &MyPlug.extra_attributes/1
```

In this example, the `:user_id` is retrieved from `conn.assigns.user.user_id` and added to the log if it exists. In the example, any values that are `nil` are filtered from the map. It is a requirement that the value is serialiazable as JSON by the Poison library, otherwise an error will be raised when attempting to encode the value.

## Log Verbosity

`LoggerJSON` plug supports two levels of logging:

  * `info` / `error` will log:

    * api_version,
    * date_time,
    * duration,
    * log_type,
    * method,
    * path,
    * request_id,
    * status

  * `warn` / `debug` will log everything from info and:

    * client_ip,
    * client_version,
    * params / request_body.

The above are default. It is possible to override them by setting a `include_debug_logging` option to:

  * `false` – means the extra debug fields (client_ip, client_version, and params) WILL NOT get logged.
  * `true` – means the extra fields WILL get logged.
  * Not setting this option will keep the defaults above.

Example:

```elixir
plug Plug.LoggerJSON,
  log: Logger.level,
  include_debug_logging: true
```

## Contributing

Before submitting your pull request, please run:

  * `mix credo --strict`,
  * `mix coveralls`,
  * `mix dialyzer`,
  *  Update changelog.

Please squash your pull request's commits into a single commit with a message and detailed description explaining the commit.

## Copyright and License

Copyright (c) 2016 Bleacher Report

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## v0.7.0 - 2019-04-22
* Updated Readme
* Updated Dependencies
* Ran Formatter
* Fixed warnings up to 1.8
* Support nested param filtering

## v0.6.0 - 2017-12-26
* Extra configuration that allows control of logging debug fields.
* Allow built-in logging keys to be suppressed
* Add Elixir 1.5.3 official support
* Add Erlang/OTP 20.1 official support
* Update dev & test dependencies

## v0.5.0 - 2017-08-21
* Extra attributes configuration that allows for logging custom metrics
* Add Elixir 1.5.1
* Add Erlang/OTP 20
* Add Erlang/OTP 19.3 to test suite

## v0.4.0 - 2017-03-18
* Breaking Changes!
* Log errors with logger level `error` instead of `info`
* Change `status` field to an integer type instead of string
* Drop official support of Elixir 1.2 and OTP 18
* Add support for Elixir 1.4 and OTP 19.1 & 19.2
* Update dependencies
* Fix dialyzer warnings

## v0.3.1 - 2017-01-05
* Update dependencies
* Fix Elixir 1.4 warnings

## v0.3.0 - 2016-11-22
* Breaking Changes!
* Drop the `fastly_duration` log value
* Add the ability to change verbosity. Log levels warn/debug will return everything from 0.2 minus fastly_duration. Log levels info/error will return a subset of warn/debug that is missing params, client_ip, & client_version.

## v0.2.0 - 2016-10-26
* Add support for error logging tagged with the request id. Errors now are bunched up as a single JSON message in addition to the standard output for easier parsing of errors and matching requests to the resulting error.


test/plug/logger_json_test.exs
defmodule Plug.LoggerJSONTest do
  use ExUnit.Case
  use Plug.Test

  import ExUnit.CaptureIO
  require Logger

  defmodule MyDebugPlug do
    use Plug.Builder

    plug(Plug.LoggerJSON, log: :debug, extra_attributes_fn: &__MODULE__.extra_attributes/1)

    plug(Plug.Parsers,
      parsers: [:urlencoded, :multipart, :json],
      pass: ["*/*"],
      json_decoder: Poison
    )

    plug(:passthrough)

    defp passthrough(conn, _) do
      Plug.Conn.send_resp(conn, 200, "Passthrough")
    end

    def extra_attributes(conn) do
      map = %{
        "user_id" => get_in(conn.assigns, [:user, :user_id]),
        "other_id" => get_in(conn.private, [:private_resource, :id]),
        "should_not_appear" => conn.private[:does_not_exist]
      }

      map
      |> Enum.filter(&(&1 !== nil))
      |> Enum.into(%{})
    end
  end

  defmodule MyInfoPlug do
    use Plug.Builder

    plug(Plug.LoggerJSON, log: :info)

    plug(Plug.Parsers,
      parsers: [:urlencoded, :multipart, :json],
      pass: ["*/*"],
      json_decoder: Poison
    )

    plug(:passthrough)

    defp passthrough(conn, _) do
      Plug.Conn.send_resp(conn, 200, "Passthrough")
    end
  end

  defmodule MyInfoPlugWithIncludeDebugLogging do
    use Plug.Builder

    plug(Plug.LoggerJSON, log: :info, include_debug_logging: true)

    plug(Plug.Parsers,
      parsers: [:urlencoded, :multipart, :json],
      pass: ["*/*"],
      json_decoder: Poison
    )

    plug(:passthrough)

    defp passthrough(conn, _) do
      Plug.Conn.send_resp(conn, 200, "Passthrough")
    end
  end

  defp remove_colors(message) do
    message
    |> String.replace("\e[36m", "")
    |> String.replace("\e[31m", "")
    |> String.replace("\e[22m", "")
    |> String.replace("\n\e[0m", "")
    |> String.replace("{\"requ", "{\"requ")
  end

  defp call(conn, plug \\ MyDebugPlug) do
    get_log(fn -> plug.call(conn, []) end)
  end

  defp get_log(func) do
    data =
      capture_io(:user, fn ->
        Process.put(:get_log, func.())
        Logger.flush()
      end)

    {Process.get(:get_log), data}
  end

  setup do
    Application.put_env(:plug_logger_json, :json_encoder, Jason)
  end

  test "correct output - no params or headers" do
    {_conn, message} =
      conn(:get, "/")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["api_version"] == "N/A"
    assert map["client_ip"] == "N/A"
    assert map["client_version"] == "N/A"
    assert map["date_time"]
    assert map["duration"]
    assert map["handler"] == "N/A"
    assert map["log_type"] == "http"
    assert map["method"] == "GET"
    assert map["params"] == %{}
    assert map["path"] == "/"
    assert map["request_id"] == nil
    assert map["status"] == 200
  end

  test "correct output - params and headers" do
    {_conn, message} =
      conn(:get, "/", fake_param: "1")
      |> put_req_header("authorization", "f3443890-6683-4a25-8094-f23cf10b72d0")
      |> put_req_header("content-type", "application/json")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["api_version"] == "N/A"
    assert map["client_ip"] == "N/A"
    assert map["client_version"] == "N/A"
    assert map["date_time"]
    assert map["duration"]
    assert map["handler"] == "N/A"
    assert map["log_type"] == "http"
    assert map["method"] == "GET"
    assert map["params"] == %{"fake_param" => "1"}
    assert map["path"] == "/"
    assert map["request_id"] == nil
    assert map["status"] == 200
  end

  test "doesn't include debug log lines for MyInfoPlug" do
    {_conn, message} =
      conn(:get, "/", fake_param: "1")
      |> put_req_header("x-forwarded-for", "209.49.75.165")
      |> put_req_header("x-client-version", "ios/1.5.4")
      |> call(MyInfoPlug)

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["client_ip"] == nil
    assert map["client_version"] == nil
    assert map["params"] == nil
  end

  test "include debug log lines for MyInfoPlugWithIncludeDebugLogging" do
    {_conn, message} =
      conn(:get, "/", fake_param: "1")
      |> put_req_header("x-forwarded-for", "209.49.75.165")
      |> put_req_header("x-client-version", "ios/1.5.4")
      |> call(MyInfoPlugWithIncludeDebugLogging)

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["client_ip"] == "209.49.75.165"
    assert map["client_version"] == "ios/1.5.4"
    assert map["params"] == %{"fake_param" => "1"}
  end

  test "correct output - Phoenix" do
    {_conn, message} =
      conn(:get, "/")
      |> put_private(:phoenix_controller, Plug.LoggerJSONTest)
      |> put_private(:phoenix_action, :show)
      |> put_private(:phoenix_format, "json")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["api_version"] == "N/A"
    assert map["client_ip"] == "N/A"
    assert map["client_version"] == "N/A"
    assert map["date_time"]
    assert map["duration"]
    assert map["handler"] == "Elixir.Plug.LoggerJSONTest#show"
    assert map["log_type"] == "http"
    assert map["method"] == "GET"
    assert map["params"] == %{}
    assert map["path"] == "/"
    assert map["request_id"] == nil
    assert map["status"] == 200
  end

  test "correct output - Post request JSON" do
    json =
      %{
        "reaction" => %{
          "reaction" => "other",
          "track_id" => "7550",
          "type" => "emoji",
          "user_id" => "a2e684ee-2e5f-4e4d-879a-bb253908eef3"
        }
      }
      |> Poison.encode!()

    {_conn, message} =
      conn(:post, "/", json)
      |> put_req_header("content-type", "application/json")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["api_version"] == "N/A"
    assert map["client_ip"] == "N/A"
    assert map["client_version"] == "N/A"
    assert map["date_time"]
    assert map["duration"]
    assert map["handler"] == "N/A"
    assert map["log_type"] == "http"
    assert map["method"] == "POST"

    assert map["params"] == %{
             "reaction" => %{
               "reaction" => "other",
               "track_id" => "7550",
               "type" => "emoji",
               "user_id" => "a2e684ee-2e5f-4e4d-879a-bb253908eef3"
             }
           }

    assert map["path"] == "/"
    assert map["request_id"] == nil
    assert map["status"] == 200
  end

  test "correct output - X-forwarded-for header" do
    {_conn, message} =
      conn(:get, "/")
      |> put_req_header("x-forwarded-for", "209.49.75.165")
      |> put_private(:phoenix_controller, Plug.LoggerJSONTest)
      |> put_private(:phoenix_action, :show)
      |> put_private(:phoenix_format, "json")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["api_version"] == "N/A"
    assert map["client_ip"] == "209.49.75.165"
    assert map["client_version"] == "N/A"
    assert map["date_time"]
    assert map["duration"]
    assert map["handler"] == "Elixir.Plug.LoggerJSONTest#show"
    assert map["log_type"] == "http"
    assert map["method"] == "GET"
    assert map["params"] == %{}
    assert map["path"] == "/"
    assert map["request_id"] == nil
    assert map["status"] == 200
  end

  test "correct output - client version header" do
    {_conn, message} =
      conn(:get, "/")
      |> put_req_header("x-client-version", "ios/1.5.4")
      |> call

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["client_version"] == "ios/1.5.4"
  end

  test "correct output - custom paths" do
    {_conn, message} =
      conn(:get, "/")
      |> put_req_header("x-client-version", "ios/1.5.4")
      |> assign(:user, %{user_id: "123"})
      |> put_private(:private_resource, %{id: 456})
      |> call()

    map =
      message
      |> remove_colors
      |> Poison.decode!()

    assert map["user_id"] == "123"
    assert map["other_id"] == 456
    refute map["should_not_appear"]
  end

  test "correct output - nested filtered keys" do
    Application.put_env(:plug_logger_json, :filtered_keys, ["password"])

    user =
      conn(:post, "/", %{user: %{password: "secret", username: "me"}})
      |> call()
      |> elem(1)
      |> remove_colors()
      |> Poison.decode!()
      |> get_in(["params", "user"])

    assert user["password"] == "[FILTERED]"
    assert user["username"] == "me"
  end

  test "correct output - structs in params" do
    params =
      conn(:post, "/", %{photo: %Plug.Upload{}})
      |> call()
      |> elem(1)
      |> remove_colors()
      |> Poison.decode!()
      |> get_in(["params"])

    assert params["photo"] == %{"content_type" => nil, "filename" => nil, "path" => nil}
  end

  describe "500 error" do
    test "logs the error" do
      stacktrace = [
        {MyDebugPlug, :index, 2, [file: 'web/controllers/reaction_controller.ex', line: 53]},
        {MyDebugPlug, :action, 2, [file: 'web/controllers/reaction_controller.ex', line: 1]},
        {MyDebugPlug, :phoenix_controller_pipeline, 2, [file: 'web/controllers/reaction_controller.ex', line: 1]},
        {MyDebugPlug, :instrument, 4, [file: 'lib/reactions/endpoint.ex', line: 1]},
        {MyDebugPlug, :dispatch, 2, [file: 'lib/phoenix/router.ex', line: 261]},
        {MyDebugPlug, :do_call, 2, [file: 'web/router.ex', line: 1]},
        {MyDebugPlug, :call, 2, [file: 'lib/plug/error_handler.ex', line: 64]},
        {MyDebugPlug, :phoenix_pipeline, 1, [file: 'lib/reactions/endpoint.ex', line: 1]},
        {MyDebugPlug, :call, 2, [file: 'lib/reactions/endpoint.ex', line: 1]},
        {Plug.Adapters.Cowboy.Handler, :upgrade, 4, [file: 'lib/plug/adapters/cowboy/handler.ex', line: 15]},
        {:cowboy_protocol, :execute, 4, [file: 'src/cowboy_protocol.erl', line: 442]}
      ]

      {_conn, _} =
        conn(:get, "/")
        |> call

      {_, message} = get_log(fn -> Plug.LoggerJSON.log_error(:error, %RuntimeError{message: "ERROR"}, stacktrace) end)

      error_log =
        message
        |> remove_colors
        |> Poison.decode!()

      assert error_log["log_type"] == "error"
      assert error_log["message"]
      assert error_log["request_id"] == nil
    end
  end
end


test/plug_logger_json_test.exs
defmodule PlugLoggerJsonTest do
  use ExUnit.Case
  doctest PlugLoggerJson

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule PlugLoggerJson.Mixfile do
  use Mix.Project

  @source_url "https://github.com/bleacherreport/plug_logger_json"
  @version "0.7.0"

  def project do
    [
      app: :plug_logger_json,
      name: "Plug Logger JSON",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      elixir: "~> 1.3",
      version: @version,
      deps: deps(),
      package: package(),
      docs: docs(),
      test_coverage: [tool: ExCoveralls],
      dialyzer: [plt_add_deps: true],
      preferred_cli_env: [
        coveralls: :test,
        "coveralls.detail": :test,
        "coveralls.post": :test,
        "coveralls.html": :test
      ]
    ]
  end

  def application do
    [
      applications: [:logger, :plug]
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.0.2", only: [:dev]},
      {:dialyxir, "~> 0.5.1", only: [:dev]},
      {:earmark, "~> 1.3.1", only: [:dev]},
      {:ex_doc, ">= 0.0.0", only: [:dev], runtime: false},
      {:excoveralls, "~> 0.10.5", only: [:test]},
      {:plug, "~> 1.0"},
      {:poison, "~> 1.5 or ~> 2.0 or ~> 3.0 or ~> 4.0"}
    ]
  end

  defp package do
    [
      description: "Elixir Plug that formats HTTP request logs as JSON.",
      files: ["lib", "mix.exs", "README*", "LICENSE*", "CHANGELOG*"],
      licenses: ["Apache-2.0"],
      maintainers: ["John Kelly, Ben Marx"],
      links: %{
        "Changelog" => "https://hexdocs.pm/plug_logger_json/changelog.html",
        "GitHub" => @source_url
      },
    ]
  end

  defp docs do
    [
      extras: [
        "CHANGELOG.md",
        "LICENSE": [title: "License"],
        "README.md": [title: "Overview"]
      ],
      main: "readme",
      source_url: @source_url,
      source_ref: @version,
      api_reference: false,
      formatters: ["html"]
    ]
  end
end


.credo.exs
# This file contains the configuration for Credo.
#
# If you find anything wrong or unclear in this file, please report an
# issue on GitHub: https://github.com/rrrene/credo/issues
%{
  #
  # You can have as many configs as you like in the `configs:` field.
  configs: [
    %{
      #
      # Run any config using `mix credo -C <name>`. If no config name is given
      # "default" is used.
      name: "default",
      #
      # these are the files included in the analysis
      files: %{
        #
        # you can give explicit globs or simply directories
        # in the latter case `**/*.{ex,exs}` will be used
        included: ["lib/", "src/", "web/", "apps/"],
        excluded: []
      },
      #
      # The `checks:` field contains all the checks that are run. You can
      # customize the parameters of any given check by adding a second element
      # to the tuple.
      #
      # There are two ways of deactivating a check:
      # 1. deleting the check from this list
      # 2. putting `false` as second element (to quickly "comment it out"):
      #
      #      {Credo.Check.Consistency.ExceptionNames, false}
      #
      checks: [
        {Credo.Check.Consistency.ExceptionNames},
        {Credo.Check.Consistency.LineEndings},
        {Credo.Check.Consistency.SpaceAroundOperators},
        {Credo.Check.Consistency.SpaceInParentheses},
        {Credo.Check.Consistency.TabsOrSpaces},

        # For some checks, like AliasUsage, you can only customize the priority
        # Priority values are: `low, normal, high, higher`
        # {Credo.Check.Design.AliasUsage, priority: :low},
        # For others you can set parameters
        {Credo.Check.Design.AliasUsage, false},
        {Credo.Check.Design.DuplicatedCode, mass_threshold: 16, nodes_threshold: 2},

        # You can also customize the exit_status of each check.
        # If you don't want TODO comments to cause `mix credo` to fail, just
        # set this value to 0 (zero).
        {Credo.Check.Design.TagTODO, exit_status: 2},
        {Credo.Check.Design.TagFIXME},

        {Credo.Check.Readability.FunctionNames},
        {Credo.Check.Readability.MaxLineLength, priority: :low, max_length: 100},
        {Credo.Check.Readability.ModuleAttributeNames},
        {Credo.Check.Readability.ModuleDoc},
        {Credo.Check.Readability.ModuleNames},
        {Credo.Check.Readability.PredicateFunctionNames},
        {Credo.Check.Readability.TrailingBlankLine},
        {Credo.Check.Readability.TrailingWhiteSpace},
        {Credo.Check.Readability.VariableNames},

        {Credo.Check.Refactor.ABCSize},
        {Credo.Check.Refactor.CaseTrivialMatches},
        {Credo.Check.Refactor.CondStatements},
        {Credo.Check.Refactor.FunctionArity},
        {Credo.Check.Refactor.MatchInCondition},
        {Credo.Check.Refactor.PipeChainStart},
        {Credo.Check.Refactor.CyclomaticComplexity},
        {Credo.Check.Refactor.NegatedConditionsInUnless},
        {Credo.Check.Refactor.NegatedConditionsWithElse},
        {Credo.Check.Refactor.Nesting},
        {Credo.Check.Refactor.UnlessWithElse},

        {Credo.Check.Warning.IExPry},
        {Credo.Check.Warning.IoInspect},
        {Credo.Check.Warning.NameRedeclarationByAssignment},
        {Credo.Check.Warning.NameRedeclarationByCase},
        {Credo.Check.Warning.NameRedeclarationByDef},
        {Credo.Check.Warning.NameRedeclarationByFn},
        {Credo.Check.Warning.OperationOnSameValues},
        {Credo.Check.Warning.UnusedEnumOperation},
        {Credo.Check.Warning.UnusedKeywordOperation},
        {Credo.Check.Warning.UnusedListOperation},
        {Credo.Check.Warning.UnusedStringOperation},
        {Credo.Check.Warning.UnusedTupleOperation},
        {Credo.Check.Warning.OperationWithConstantResult},
      ]
    }
  ]
}


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :plug_logger_json, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:plug_logger_json, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"
#
config :logger, :console,
  format: "$message\n",
  level: :debug,
  metadata: [:request_id]

config :plug_logger_json,
  filtered_keys: ["password", "authorization"]


lib/plug/logger_json.ex
defmodule Plug.LoggerJSON do
  @moduledoc """
  A plug for logging basic request information in the format:

  ```json
  {
    "api_version":     "N/A"
    "client_ip":       "23.235.46.37"
    "client_version":  "ios/1.6.7",
    "date_time":       "2016-05-31T18:00:13Z",
    "duration":        4.670,
    "handler":         "fronts#index"
    "log_type":        "http",
    "method":          "POST",
    "params":          {
                         "user":"jkelly",
                         "password":"[FILTERED]"
                       },
    "path":            "/",
    "request_id":      "d90jcl66vp09r8tke3utjsd1pjrg4ln8",
    "status":          "200"
  }
  ```

  To use it, just plug it into the desired module.

      plug Plug.LoggerJSON, log: :debug

  ## Options

  * `:log` - The log level at which this plug should log its request info.
    Default is `:info`.

  * `:extra_attributes_fn` - Function to call with `conn` to add additional
    fields to the requests. Default is `nil`. Please see "Extra Fields" section
    for more information.

  ## Extra Fields

  Additional data can be logged alongside the request by specifying a function
  to call which returns a map:

      def extra_attributes(conn) do
        map = %{
          "user_id" => get_in(conn.assigns, [:user, :user_id]),
          "other_id" => get_in(conn.private, [:private_resource, :id]),
          "should_not_appear" => conn.private[:does_not_exist]
        }

        map
        |> Enum.filter(&(&1 !== nil))
        |> Enum.into(%{})
      end

      plug Plug.LoggerJSON, log: Logger.level,
                            extra_attributes_fn: &MyPlug.extra_attributes/1

  In this example, the `:user_id` is retrieved from `conn.assigns.user.user_id`
  and added to the log if it exists. In the example, any values that are `nil`
  are filtered from the map. It is a requirement that the value is
  serialiazable as JSON by the Poison library, otherwise an error will be raised
  when attempting to encode the value.
  """

  alias Plug.Conn

  @behaviour Plug

  require Logger

  @typedoc """
  Type for a plug option
  """
  @type opts :: binary | tuple | atom | integer | float | [opts] | %{opts => opts}

  @typedoc """
  Type for time
  """
  @type time :: {non_neg_integer(), non_neg_integer(), non_neg_integer()}

  @spec init(opts) :: opts
  def init(opts), do: opts

  @spec call(Plug.Conn.t(), opts) :: Plug.Conn.t()
  def call(conn, level_or_opts) when is_atom(level_or_opts) do
    call(conn, level: level_or_opts)
  end

  def call(conn, opts) do
    level = Keyword.get(opts, :log, :info)
    start = :os.timestamp()

    Conn.register_before_send(conn, fn conn ->
      :ok = log(conn, level, start, opts)
      conn
    end)
  end

  @spec log(Plug.Conn.t(), atom(), time(), opts) :: atom() | no_return()
  def log(conn, level, start, opts \\ [])
  def log(conn, :error, start, opts), do: log(conn, :info, start, opts)
  def log(conn, :info, start, opts), do: log_message(conn, :info, start, opts)
  def log(conn, :warn, start, opts), do: log(conn, :debug, start, opts)

  def log(conn, :debug, start, opts) do
    log_message(conn, :info, start, Keyword.put_new(opts, :include_debug_logging, true))
  end

  @spec log_error(atom(), map(), list()) :: atom()
  def log_error(kind, reason, stacktrace) do
    _ =
      Logger.log(:error, fn ->
        %{
          "log_type" => "error",
          "message" => Exception.format(kind, reason, stacktrace),
          "request_id" => Logger.metadata()[:request_id]
        }
        |> Poison.encode!()
      end)
  end

  @spec log_message(Plug.Conn.t(), atom(), time(), opts) :: atom()
  defp log_message(conn, level, start, opts) do
    Logger.log(level, fn ->
      conn
      |> basic_logging(start)
      |> Map.merge(debug_logging(conn, opts))
      |> Map.merge(phoenix_attributes(conn))
      |> Map.merge(extra_attributes(conn, opts))
      |> Poison.encode!()
    end)
  end

  defp basic_logging(conn, start) do
    stop = :os.timestamp()
    duration = :timer.now_diff(stop, start)
    req_id = Logger.metadata()[:request_id]
    req_headers = format_map_list(conn.req_headers)

    log_json = %{
      "api_version" => Map.get(req_headers, "accept", "N/A"),
      "date_time" => iso8601(:calendar.now_to_datetime(:os.timestamp())),
      "duration" => Float.round(duration / 1000, 3),
      "log_type" => "http",
      "method" => conn.method,
      "path" => conn.request_path,
      "request_id" => req_id,
      "status" => conn.status
    }

    Map.drop(log_json, Application.get_env(:plug_logger_json, :suppressed_keys, []))
  end

  defp extra_attributes(conn, opts) do
    case Keyword.get(opts, :extra_attributes_fn) do
      fun when is_function(fun) -> fun.(conn)
      _ -> %{}
    end
  end

  @spec client_version(%{String.t() => String.t()}) :: String.t()
  defp client_version(headers) do
    headers
    |> Map.get("x-client-version", "N/A")
    |> case do
      "N/A" ->
        Map.get(headers, "user-agent", "N/A")

      accept_value ->
        accept_value
    end
  end

  @spec debug_logging(Plug.Conn.t(), opts) :: map()
  defp debug_logging(conn, opts) do
    case Keyword.get(opts, :include_debug_logging) do
      true ->
        req_headers = format_map_list(conn.req_headers)

        %{
          "client_ip" => format_ip(Map.get(req_headers, "x-forwarded-for", "N/A")),
          "client_version" => client_version(req_headers),
          "params" => format_map_list(conn.params)
        }

      _ ->
        %{}
    end
  end

  @spec filter_values(struct(), [binary()]) :: binary()
  defp filter_values(%{__struct__: mod} = struct, filters) when is_atom(mod) do
    struct
    |> Map.from_struct()
    |> filter_values(filters)
  end

  @spec filter_values(map(), [binary()]) :: [{binary(), any()}]
  defp filter_values(%{} = map, filters) do
    Enum.into(map, %{}, fn {k, v} ->
      if is_binary(k) and k in filters do
        {k, "[FILTERED]"}
      else
        {k, filter_values(v, filters)}
      end
    end)
  end

  @spec filter_values([{binary(), any()}], [binary()]) :: [{binary(), any()}]
  defp filter_values(list, filters) when is_list(list) do
    Enum.map(list, &filter_values(&1, filters))
  end

  defp filter_values(other, _filters), do: format_value(other)

  @spec format_ip(String.t()) :: String.t()
  defp format_ip("N/A") do
    "N/A"
  end

  defp format_ip(x_forwarded_for) do
    hd(String.split(x_forwarded_for, ", "))
  end

  @spec format_map_list(Enumerable.t()) :: map()
  defp format_map_list(enumerable) do
    enumerable
    |> filter_values(Application.get_env(:plug_logger_json, :filtered_keys, []))
    |> Enum.into(%{})
  end

  defp format_value(value) when is_binary(value) do
    String.slice(value, 0..500)
  end

  defp format_value(value) do
    value
  end

  defp iso8601({{year, month, day}, {hour, minute, second}}) do
    zero_pad(year, 4) <>
      "-" <>
      zero_pad(month, 2) <>
      "-" <>
      zero_pad(day, 2) <>
      "T" <>
      zero_pad(hour, 2) <>
      ":" <>
      zero_pad(minute, 2) <>
      ":" <>
      zero_pad(second, 2) <>
      "Z"
  end

  @spec phoenix_attributes(map()) :: map()
  defp phoenix_attributes(%{private: %{phoenix_controller: controller, phoenix_action: action}}) do
    %{"handler" => "#{controller}##{action}"}
  end

  defp phoenix_attributes(_) do
    %{"handler" => "N/A"}
  end

  @spec zero_pad(1..3_000, non_neg_integer()) :: String.t()
  defp zero_pad(val, count) do
    num = Integer.to_string(val)
    :binary.copy("0", count - byte_size(num)) <> num
  end
end
